notes/force-move-implementation.md

```diff
@@ -164,9 +175,110 @@ Checking signatures is likely to be a signification part of the logic. Here's a
 
 **Can we optimize this further by changing the way states / sigs are passed in?** For example passing the arrays in last-state-first might help.
 
-## Updates call signatures
+## ForceMove interface
+
+With these considerations in mind, the ForceMove implementation should be something like the below.
+
+### Types and storage
+
+```javascript
+
+struct FixedPart {
+ string chainId;
+ address[] participants;
+ uint256 channelNonce;
+}
+
+struct VariablePart {
+  uint256 turnNum;
+  bool isFinal;
+}
+
+struct State { // participants sign this
+  uint256 turnNum;
+  bool isFinal;
+  bytes32 channelId;  // keccack(FixedPart)
+  bytes32 variablePartHash; //keccak(VariablePart)
+}
+
+struct ChannelStorage {
+  uint256 turnNumRecord;
+  uint256 finalizesAt;
+  bytes32 stateHash; // keccak(State)
+  address challengerAddress;
+}
+
+mapping(address => bytes32) public channelStorageHashes;
 
-[TODO] Rewrite the method call signatures to avoid any unnecessary repetition in the data being passed in.
+```
+
+### Public methods:
+
+```javascript
+
+
+function forceMove(uint turnNumRecord, FixedPart fixedPart, VariableParts[] variableParts, uint newTurnNumRecord, bool[] isFinals, Signature[] sigs, Signature challengerSig) public returns() {
+
+  address channelID = keccak(fixedPart);
+
+  // ------------
+  // REQUIREMENTS
+  // ------------
+
+  require(keccack(ChannelStorage(turnNumRecord, 0, 0, 0)) == channelStorageHashes[channelId],'Channel not open')
+  require(_validNChain(variableParts, newTurnNumRecord, isFinals),'Not a valid chain of n commitments');
```

Comment:
Currently this needs the `fixedPart` too, as we pass the entire state through to the app's `validTransition` function. It's not clear whether this is required though, or whether passing just the `outcome` and `appData` would suffice. This is missing the `appDefinition`, which it definitely needs.

Reply:
In 4de94bf I am passing the `fixedPart` (which now includes `appDefinition`) all the way down to `validTransition`.